### PR TITLE
adding observer to query

### PR DESCRIPTION
**Pull request by JavierZunzunegui**
Introducing `QueryObserver` and adding it to `Query` which gets called every time the query is executed (via `ExecutableQuery.attempt`, modified slightly).

Introduced to allow monitoring of gocql, should be generic enough to support all monitoring libraries.

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -1584,6 +1616,31 @@ func (t *traceWriter) Trace(traceId []byte) {
...
+}
+
+// TODO - BatchObservation, similar to QueryObservation but adapted to Batch
+
+// QueryObserver is the interface implemented by query observers / stat collectors.
+type QueryObserver interface {
+	// Observe gets called on every query to cassandra, including all queries in an iterator when paging is enabled.
+	// It doesn't get called if there is no query because the session is closed or there are no connections available.
+	// The error reported only shows query errors, i.e. if a SELECT is valid but finds no matches it will be nil.
+	Observe(QueryObservation)
```
Comment (Zariel) [inline session.go]: Call this ObserveQuery instead and change the signature to take a context as the first param

File: session.go [hunk 2]
```diff
@@ -299,6 +300,14 @@ func (s *Session) SetTrace(trace Tracer) {
 	s.mu.Unlock()
 }
 
+// SetQueryObserver sets the default query-level observer for this session. This setting can also
+// be changed on a per-query basis.
+func (s *Session) SetQueryObserver(observer QueryObserver) {
```
Comment (Zariel) [inline session.go]: I eventually want to remove the option to set globals on the session to remove the lock on querying #357, please remove this part. Then the observer only need to live on the config.

File: session.go [hunk 3]
```diff
@@ -720,6 +731,13 @@ func (q *Query) Trace(trace Tracer) *Query {
 	return q
 }
 
+// QueryObserver enables query-level observer on this query.
+// The provided observer will be called every time this query is executed.
+func (q *Query) Observer(observer QueryObserver) *Query {
```
Comment (Zariel) [inline session.go]: Should also add this to Batch

File: session.go [hunk 4]
```diff
@@ -1584,6 +1616,31 @@ func (t *traceWriter) Trace(traceId []byte) {
...
 
+type QueryObservation struct {
+	ctx      context.Context // the query's context. Use it to pass arbitrary data to the collector
+	keyspace string
+	stmt     string
+
+	start time.Time // time immediately before the query was called
+	end   time.Time // time immediately after the query returned
+
+	rows int // the number of rows in the current iter. In multi-scans, rows from previous scans are not counted
```
Comment (Zariel) [inline session.go]: s/multi-scans/paginated queries/

File: cassandra_test.go [hunk 5]
```diff
@@ -184,6 +185,97 @@ func TestTracing(t *testing.T) {
...
 
+func TestObserve(t *testing.T) {
+	session := createSession(t)
+	defer session.Close()
+
+	if err := createTable(session, `CREATE TABLE gocql_test.observe (id int primary key)`); err != nil {
+		t.Fatal("create:", err)
+	}
+
+	var observedErr error
```
Comment (Zariel) [inline cassandra_test.go]: declare these in a block

File: session.go [hunk 6]
```diff
@@ -770,10 +788,22 @@ func (q *Query) execute(conn *Conn) *Iter {
...
 	// TODO: track latencies per host and things as well instead of just total
+
+	if q.observer != nil {
+		q.observer.Observe(QueryObservation{
+			ctx:      q.context,
+			keyspace: q.session.pool.keyspace,
+			stmt:     q.stmt,
+			start:    start,
+			end:      end,
+			rows:     iter.numRows - iter.pos,
```
Comment (Zariel) [inline session.go]: iter.pos here should always be 0, is this what you intend?

Comment (JavierZunzunegui) [reply session.go]: Would it not be 0 in paginated queries?

File: session.go (hunk 3 above)
Comment (JavierZunzunegui) [reply session.go]: Part of a slightly larger question - how to observe Batches? I'm not familiar with those, but from what I can see the `QueryObserver` may be called for each statement in the batch (with shared times and error) although I don't know if the rows have any meaning in batch queries.

File: session.go (hunk 6 above)
Comment (JavierZunzunegui) [reply session.go]: yeah, always 0. Made TestObserve_Pagination to check it

**Comment by JavierZunzunegui** [comment]
@Zariel any further feedback on this?

File: session.go (hunk 3 above)
Comment (Zariel) [reply session.go]: Batch will have multiple statements but no return values unless BatchCAS is used.

**Comment by JavierZunzunegui** [comment]
@Zariel added observe support for Batches

**Comment by Zariel** [comment]
I would like somehow to have batches come through as a single observation as they resulted from a single query that went to a single host.

File: session.go [hunk 7]
```diff
@@ -1584,6 +1628,32 @@ func (t *traceWriter) Trace(traceId []byte) {
 	}
 }
 
+type ObserveQuery struct {
```
Comment (Zariel) [inline session.go]: I had meant this the other way around, this is fine to be called `ObservedQuery`

File: session.go [hunk 8]
```diff
@@ -1584,6 +1628,32 @@ func (t *traceWriter) Trace(traceId []byte) {
...
+	Err error
+}
+
+// QueryObserver is the interface implemented by query observers / stat collectors.
+type QueryObserver interface {
+	// Observe gets called on every query to cassandra, including all queries in an iterator when paging is enabled.
+	// It also gets called once for each query in a batch.
+	// It doesn't get called if there is no query because the session is closed or there are no connections available.
+	// The error reported only shows query errors, i.e. if a SELECT is valid but finds no matches it will be nil.
+	Observe(context.Context, ObserveQuery)
```
Comment (Zariel) [inline session.go]: But name this method `ObserveQuery`

File: session.go [hunk 9]
```diff
@@ -1445,10 +1476,23 @@ func (b *Batch) WithTimestamp(timestamp int64) *Batch {
...
-func (b *Batch) attempt(d time.Duration) {
+func (b *Batch) attempt(keyspace string, end, start time.Time, iter *Iter) {
 	b.attempts++
-	b.totalLatency += d.Nanoseconds()
+	b.totalLatency += end.Sub(start).Nanoseconds()
 	// TODO: track latencies per host and things as well instead of just total
+
+	if b.observer != nil {
+		for _, entry := range b.Entries {
+			b.observer.Observe(b.context, ObserveQuery{
```
Comment (Zariel) [inline session.go]: I think we should here only get a single observation for a batch

File: session.go [hunk 10]
```diff
@@ -1584,6 +1628,32 @@ func (t *traceWriter) Trace(traceId []byte) {
 	}
 }
 
+type ObserveQuery struct {
+	Keyspace string
+	Stmt     string
```
Comment (Zariel) [inline session.go]: call this `Statement`

**Comment by Zariel** [comment]
Would It make more sense to do something akin to what `net/http` does? https://golang.org/pkg/net/http/httptrace/

**Comment by zllak** [comment]
We've done something very similar on a fork. We are running this in production to export metrics to prometheus. It works very well and it would be awesome to have this integrated into gocql directly !

**Comment by JavierZunzunegui** [comment]
@Zariel the monitoring approach in https://golang.org/pkg/net/http/httptrace/ relies in context. I don't think gocql should emulate that because:
- gocql already uses the non-context approach for tracing, it would be inconsistent to use content for monitoring.
- the context approach is, in my opinion, slightly lower level in that the caller is responsible for setting the monitoring methods in the context **for every request** as by default it is nil. The approach here sets a default value which is much more useful if you want all queries monitored by default and in a similar fashion. Of course you could do the same in the context approach wrapping the gocql session with something else that sets the observer in each given context, but that is extra complexity for no gain

**Comment by Zariel** [comment]
Can you add that the interface is experimental then I’m happy to merge.
Thanks very much

**Comment by Zariel** [comment]
@JavierZunzunegui something to note about tracing in Gocql is that is for tracing on the Cassandra side, which tells cassandra that we want to trace the query and allows us to read the results.

**Comment by Zariel** [comment]
@zllak could you share a bit more about how you use it? Is the current interface sufficient in Gocql now?

**Comment by zllak** [comment]
The interface provided here is what we (almost) exactly did. We use this to create some prometheus probe, to graph latencies of calls, count the number of select/update/delete (regexp on the statement tho :/) per table/keyspace. I'll update on our side to use this new interface and see, but it looks very good !

**Comment by zllak** [comment]
@Zariel the only difference I can see is that it only report network errors, not "logical" errors like Notfound, or context.Canceled. We use those one. I'll try it tomorrow.

**Pull request merged**